src/storage/storage_backend.rs

```diff
@@ -164,6 +165,14 @@ pub trait StorageBackend<User: UserDetail>: Send + Sync + Debug {
     /// The concrete type of the _metadata_ used by this storage backend.
     type Metadata: Metadata + Sync + Send;
 
+    /// Restrict the backend's capabilities so that it may only access files underneath `path`.
+    /// Once restricted, it may never be unrestricted.
+    ///
+    /// The path should be absolute.
+    fn enter<P: AsRef<Path>>(&mut self, _path: P) -> io::Result<()> {
+        Err(io::Error::new(io::ErrorKind::Other, "Not supported by this StorageBackend"))
```

Comment:
Maybe it can work if its not named `enter` but rather a more generic `back-end pre-use` kind of method that will be called once for a storage back-end before its used. In that way most will have a no-op implementation or it can be use for other things too.

Not sure will have to think a bit.

Reply:
Creating a separate trait would probably require downcasting, which is ugly IMHO.  And BTW it's possible to implement `enter` without capsicum.  The `enter` method just descends into a user's home directory with no security guarantees, but the `capsicum::enter` part is what turns on enforcement.

Reply:
ok lets leave it `enter` but let the default be a no-op instead of returning an error.

Reply:
Actually, lets make it a bit more generic:

```rust
    /// Called to prepare access to the back-end for the user.
    /// 
    /// Implementations may for instance restrict the backend's capabilities so that it may 
    /// only access files underneath a user's home directory.
    ///
    /// The path should be absolute.
    fn enter<P: AsRef<Path>>(&mut self, user: &User) -> io::Result<()> {
        Ok(())
    }
```

Reply:
Changing the default to a no-op won't have anything to do with OS-portability.  Instead, that will just make it possible to use the method on other storage backends that might share an authenticator..  That's what you want, right?  Currently, the server will never call this method unless `UserDetail::home` method returns something.

Reply:
If we make it more generic, then the implementation can restrict more things than just the path.

For instance given a `UserDetail` implementation like we have in unFTP:

```
/// The unFTP user details
#[derive(Debug, PartialEq, Eq)]
pub struct User {
    pub username: String,
    pub name: Option<String>,
    pub surname: Option<String>,
    /// Tells whether this user can log in or not.
    pub account_enabled: bool,
    /// What FTP commands can the user perform
    pub vfs_permissions: VfsOperations,
    /// For some users we know they will only upload a certain type of file
    pub allowed_mime_types: Option<Vec<String>>, // TODO: Look at https://crates.io/crates/infer to do this
    /// The user's home directory relative to the storage back-end root
    pub root: Option<PathBuf>,
}
```

The implementation can also restrict read or write operations properly with sandboxing methods.

Reply:
> Currently, the server will never call this method unless UserDetail::home method returns something.
I suggested 2 changes for this.